core/src/main/java/org/infinispan/util/concurrent/CompletionStages.java

```diff
@@ -144,20 +144,19 @@ public static boolean isCompletedSuccessfully(CompletionStage<?> stage) {
          }
          return delay;
       }
-      CompletableFuture<V> returnedFuture = new CompletableFuture<>();
-      delay.whenCompleteAsync((v, t) -> {
-         if (t != null) {
-            if (trace) {
-               log.tracef("Continuing execution of id %s with exception %s", traceId, t.getMessage());
-            }
-            returnedFuture.completeExceptionally(t);
-         } else {
-            if (trace) {
-               log.tracef("Continuing execution of id %s", traceId);
-            }
-            returnedFuture.complete(v);
+      CompletionStage<V> returnedFuture = delay.thenApplyAsync(v -> {
+         if (trace) {
+            log.tracef("Continuing execution of id %s", traceId);
          }
+         return v;
       }, continuationExecutor);
+      if (trace) {
+         // We rely on chained methods on the same future are performed in the same order
+         returnedFuture.exceptionally(t -> {
```

Comment:
If we did `return returnedFuture.exceptionally(...)` we wouldn't need to rely on any implementation details

Reply:
The problem is then I have to wrap the inner exception with `CompletionException` which can change the exception that is returned.